### PR TITLE
Send correct receipts when viewing a room

### PR DESCRIPTION
**Pull request by weeman1337**
Send receipts for the `main` thread when reading the main timeline (and not an unthreaded receipt).
I have tried to split the code into several functions to make it easier to understand.
There are now also tests for the receipt cases.

fixes https://github.com/vector-im/element-web/issues/25196

PSF-2005

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Send correct receipts when viewing a room ([\#10864](https://github.com/matrix-org/matrix-react-sdk/pull/10864)). Fixes vector-im/element-web#25196.

### CODE REVIEW

**Review by dbkr — APPROVED** [review]

File: src/components/structures/TimelinePanel.tsx [hunk 1]
```diff
@@ -971,26 +971,69 @@ class TimelinePanel extends React.Component<IProps, IState> {
...
+    /**
+     * Whether a fully_read marker should be send.
+     */
+    private shouldSendRM(readMarkerEventId: string | null): readMarkerEventId is string {
+        if (!this.state.readMarkerEventId) {
+            // Nothing that can be send.
+            return false;
+        }
+
+        if (this.lastRRSentEventId && this.lastRMSentEventId === this.state.readMarkerEventId) {
```
Comment (weeman1337) [inline src/components/structures/TimelinePanel.tsx:1009]:
```suggestion
        if (this.lastRMSentEventId && this.lastRMSentEventId === this.state.readMarkerEventId) {
```

File: src/components/structures/TimelinePanel.tsx [hunk 2]
```diff
@@ -971,26 +971,69 @@ class TimelinePanel extends React.Component<IProps, IState> {
...
+            // Prevent sending the same receipt twice.
+            return false;
+        }
+
+        if (this.state.readMarkerEventId && this.state.readMarkerEventId === this.initialReadMarkerId) {
+            // The initial read marker is a historical one.
```
Comment (weeman1337) [inline src/components/structures/TimelinePanel.tsx:1015]: Improve comment that mentions it should not send the read marker from room account data again.

**Review by andybalaam — CHANGES_REQUESTED** [review]
Awesome fix, thank you.

Did an IRL review and asked for some minor fixes, plus one more test, to cover the use of the `<` sign in the code.

File: src/components/structures/TimelinePanel.tsx [hunk 3]
```diff
@@ -1003,110 +1046,121 @@ class TimelinePanel extends React.Component<IProps, IState> {
...
-                            this.lastRRSentEventId = undefined;
-                        }
-                    } else {
-                        logger.error(e);
-                    }
-                    // it failed, so allow retries next time the user is active
-                    this.lastRRSentEventId = undefined;
-                    this.lastRMSentEventId = undefined;
-                });
+        if (shouldSendRR && lastReadEvent) {
```
Comment (weeman1337) [inline src/components/structures/TimelinePanel.tsx:1105]: Add the guard if `lastReadEvent` is set to `shouldSendRR`.

File: src/components/structures/TimelinePanel.tsx [hunk 4]
```diff
@@ -971,26 +971,69 @@ class TimelinePanel extends React.Component<IProps, IState> {
...
+            return ReceiptType.Read;
+        }
 
-        let shouldSendRR = true;
+        if (
+            !(await client.doesServerSupportUnstableFeature("org.matrix.msc2285.stable")) ||
+            !(await client.isVersionSupported("v1.4"))
+        ) {
+            // The server does not support private read receipt. Fall back to public ones.
+            return ReceiptType.Read;
```
Comment (weeman1337) [inline src/components/structures/TimelinePanel.tsx:994]: Log this case

File: src/components/structures/TimelinePanel.tsx [hunk 5]
```diff
@@ -1003,110 +1046,121 @@ class TimelinePanel extends React.Component<IProps, IState> {
...
+
+    /**
+     * Sends a fully_read marker for readMarkerEvent.
+     * Resets the last sent event Id in case of an error, so that it will be retried next time.
+     */
+    private async sendReadMarker(client: MatrixClient, readMarkerEvent: MatrixEvent): Promise<void> {
+        this.lastRMSentEventId = this.state.readMarkerEventId;
+
+        try {
+            await client.sendReadReceipt(readMarkerEvent, ReceiptType.FullyRead, true);
```
Comment (weeman1337) [inline src/components/structures/TimelinePanel.tsx:1149]: Use `read_markers` API (it was used before)

File: src/components/structures/TimelinePanel.tsx [hunk 6]
```diff
@@ -1003,110 +1046,121 @@ class TimelinePanel extends React.Component<IProps, IState> {
...
-                    dis.dispatch({
-                        action: "on_room_read",
-                        roomId: this.props.timelineSet.room.roomId,
-                    });
-                }
+        if (shouldSendRM) {
+            const readMarkerEvent = this.props.timelineSet.findEventById(readMarkerEventId);
+
+            if (readMarkerEvent) {
+                proms.push(this.sendReadMarker(client, readMarkerEvent));
```
Comment (weeman1337) [inline src/components/structures/TimelinePanel.tsx:1113]: Add a comment that it does not make sense to send a read marker for an event that is not part of the timeline.

File: test/components/structures/TimelinePanel-test.tsx [hunk 7]
```diff
@@ -98,68 +101,195 @@ const setupTestData = (): [MatrixClient, Room, MatrixEvent[]] => {
...
+                timelineSet.addLiveEvent(threadEv1, {});
+                await flushPromises();
+
+                // @ts-ignore
+                await timelinePanel.sendReadReceipts();
+
+                // fully_read is not supported for threads per spec
+                expect(client.sendReadReceipt).not.toHaveBeenCalledWith(threadEv1, ReceiptType.FullyRead, true);
+                expect(client.sendReadReceipt).toHaveBeenCalledWith(threadEv1, ReceiptType.Read);
+            });
```
Comment (andybalaam) [inline test/components/structures/TimelinePanel-test.tsx:292]: Add a test that covers the case when the event we scrolled to is _before_ the last sent read receipt.

Comment (weeman1337) [reply test/components/structures/TimelinePanel-test.tsx:363]: See https://github.com/matrix-org/matrix-react-sdk/pull/10864/commits/2be2635e65842c5b357d0e3deb70337948a7a9c1

**Review by andybalaam — APPROVED** [review]
Great change, thank you!

**Pull request merged**